# Assignment 1



Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe.

To create the above dataframe:

*   The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
*   Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
*   More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11  in the above table.
*   If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
*   Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
*   In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe

In [139]:
import pandas as pd # library for data analsysis
import numpy as np
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
import json # library to handle JSON files
import requests # library to handle requests
from shapely.geometry import Point
from bs4 import BeautifulSoup
import xml
!pip install geopandas
import geopandas as gpd
import matplotlib.pyplot as plt
!pip install folium pandas
import folium
from sklearn.cluster import KMeans
print("Libraries installed")

Libraries installed


*Scrape the List of postal codes of Canada*


In [105]:
List_url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
resource = requests.get(List_url).text

In [106]:
soup = BeautifulSoup(resource, 'xml')

In [107]:
table=soup.find('table')

In [108]:
#dataframe will consist of three columns: PostalCode, Borough, and Neighbourhood
column_names = ['Postalcode','Borough','Neighbourhood']
df = pd.DataFrame(columns = column_names)

In [109]:
# Search all the postcode, borough, neighbourhood 
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data

In [110]:
df.head()

Postalcode           Borough              Neighbourhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

*Time to clen data and remove rows were **Borough** is 'Not assigned'*

In [111]:
df=df[df['Borough']!='Not assigned']

In [112]:
df.head(10)

Postalcode           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8         M9A         Etobicoke      Islington Avenue, Humber Valley Village
9         M1B       Scarborough                               Malvern, Rouge
11        M3B        North York                                    Don Mills
12        M4B         East York              Parkview Hill, Woodbine Gardens
13        M5B  Downtown Toronto                     Garden District, Ryerson

In [113]:
temp_df=df.groupby('Postalcode')['Neighbourhood'].apply(lambda x: "%s" % ', '.join(x))
temp_df=temp_df.reset_index(drop=False)
temp_df.rename(columns={'Neighbourhood':'Neighbourhood_joined'},inplace=True)

In [114]:
df_merge = pd.merge(df, temp_df, on='Postalcode')

In [115]:
df_merge.drop(['Neighbourhood'],axis=1,inplace=True)

In [116]:
df_merge.drop_duplicates(inplace=True)

In [117]:
df_merge.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)

In [118]:
df_merge.head()

Postalcode           Borough                         Neighbourhood_joined
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [119]:
df_merge.shape

(103, 3)

# Assignment 2

**Geting latitude and logitude data**

Using file https://cocl.us/Geospatial_data to get the data.

Read the CSV data to a pandas dataframe.

In [120]:
geospatial_data = pd.read_csv('https://cocl.us/Geospatial_data')

In [121]:
#exploring data
print("Geospatial dataframe's shape: ", geospatial_data.shape)
geospatial_data.head()

Geospatial dataframe's shape:  (103, 3)


Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476


The size (103 rows) matches that of our original dataframe df.

**Combine two dataframes**

Time to combine two dataframes so that the latitude and logitude values from 'geospatial_data' are assigned to their corresponding rows in our dataframe, 'df'.

In [122]:
# Concatenate the two dataframes using column "PostalCode" from 'df' and
# "Postal Code" from 'geospatial_data'.
df = pd.concat(
    [df.set_index('Postalcode'), geospatial_data.set_index('Postal Code')],
    axis=1, join='inner')
df.head()

Borough                                Neighbourhood   Latitude  \
M3A        North York                                    Parkwoods  43.753259   
M4A        North York                             Victoria Village  43.725882   
M5A  Downtown Toronto                    Regent Park, Harbourfront  43.654260   
M6A        North York             Lawrence Manor, Lawrence Heights  43.718518   
M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government  43.662301   

     Longitude  
M3A -79.329656  
M4A -79.315572  
M5A -79.360636  
M6A -79.464763  
M7A -79.389494

In [123]:
#Postal code has been set as the index, change it to a non-index column and reset index.
df.reset_index(inplace=True)
df.head()

index           Borough                                Neighbourhood  \
0   M3A        North York                                    Parkwoods   
1   M4A        North York                             Victoria Village   
2   M5A  Downtown Toronto                    Regent Park, Harbourfront   
3   M6A        North York             Lawrence Manor, Lawrence Heights   
4   M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [124]:
#Rename the 'index' column to 'PostalCode'
df.rename(columns={'index': 'PostalCode'}, inplace=True)
df.head(12)

PostalCode           Borough  \
0         M3A        North York   
1         M4A        North York   
2         M5A  Downtown Toronto   
3         M6A        North York   
4         M7A  Downtown Toronto   
5         M9A         Etobicoke   
6         M1B       Scarborough   
7         M3B        North York   
8         M4B         East York   
9         M5B  Downtown Toronto   
10        M6B        North York   
11        M9B         Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                           Parkwoods  43.753259 -79.329656  
1                                    Victoria Village  43.725882 -79.315572  
2                           Regent Park, Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4         Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
5             Islington Avenue, Humber Valley Village  43.667856 -79.532242  
6                                      Malvern, Rouge  43.806686 -79.194353  
7                                           Don Mills  43.745906 -79.352188  
8                     Parkview Hill, Woodbine Gardens  43.706397 -79.309937  
9                            Garden District, Ryerson  43.657162 -79.378937  
10                                          Glencairn  43.709577 -79.445073  
11  West Deane Park, Princess Gardens, Martin Grov...  43.650943 -79.554724

In [125]:
print("Dataframe shape: ", df.shape)

Dataframe shape:  (103, 5)


#Assignment 3

Explore and cluster the neighborhoods in Toronto

In [126]:
#adding coordinates
df['Coordinates'] = list(zip(df['Latitude'], df['Longitude']))

In [127]:
df.head()

PostalCode           Borough                                Neighbourhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude                               Coordinates  
0  43.753259 -79.329656                 (43.7532586, -79.3296565)  
1  43.725882 -79.315572  (43.725882299999995, -79.31557159999998)  
2  43.654260 -79.360636                 (43.6542599, -79.3606359)  
3  43.718518 -79.464763  (43.718517999999996, -79.46476329999999)  
4  43.662301 -79.389494                 (43.6623015, -79.3894938)

In [128]:
df['Coordinates'] = df['Coordinates'].apply(Point)

In [129]:
gdf = gpd.GeoDataFrame(df, geometry = 'Coordinates')

In [130]:
#droping Latitude and Longitude columns
gdf.drop(['Latitude','Longitude'], axis=1)

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                         Neighbourhood  \
0                                            Parkwoods   
1                                     Victoria Village   
2                            Regent Park, Harbourfront   
3                     Lawre

In [131]:
#working with only boroughs that contain the word Toronto 
gdf = gdf[gdf.Borough.str.contains("Toronto")]

In [132]:
gdf.head()

PostalCode           Borough                                Neighbourhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude                 Coordinates  
2   43.654260 -79.360636  POINT (43.65426 -79.36064)  
4   43.662301 -79.389494  POINT (43.66230 -79.38949)  
9   43.657162 -79.378937  POINT (43.65716 -79.37894)  
15  43.651494 -79.375418  POINT (43.65149 -79.37542)  
19  43.676357 -79.293031  POINT (43.67636 -79.29303)

In [142]:

toronto_map = folium.Map(location=[43.65, -79.4], zoom_start=12)

X = df['Latitude']
Y = df['Longitude']
Z = np.stack((X, Y), axis=1)

kmeans = KMeans(n_clusters=4, random_state=0).fit(Z)

clusters = kmeans.labels_
colors = ['red', 'green', 'blue', 'yellow']
df['Cluster'] = clusters

for latitude, longitude, borough, cluster in zip(df['Latitude'], df['Longitude'], df['Borough'], df['Cluster']):
    label = folium.Popup(borough, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='black',
        fill=True,
        fill_color=colors[cluster],
        fill_opacity=0.7).add_to(toronto_map)  

toronto_map